File: src/python/generator/lib/spannertruetime/spanner_truetime.py
import urllib.request
import urllib.error
import json
import time
import sys
import threading

class SpannerTrueTimeGenerator:
    def __init__(self):
        self.session_name = None
        self.lock = threading.Lock()

        for i in range(30):
            try:
                req = urllib.request.Request('http://spanner:9020/', method='GET')
                with urllib.request.urlopen(req) as response:
                    if response.status in (200, 404):
                        break
            except urllib.error.HTTPError as e:
                if e.code == 404:
                    break
            except Exception as e:
                print("Waiting for Spanner TrueTime emulator...", flush=True)
                time.sleep(2)

        try:
            db_name = "projects/test-project/instances/test-instance/databases/test-db"
            url = f"http://spanner:9020/v1/{db_name}/sessions"
            data = json.dumps({}).encode('utf-8')
            req = urllib.request.Request(url, data=data, headers={'Content-Type': 'application/json'}, method='POST')
            
            with urllib.request.urlopen(req) as response:
                res_data = json.loads(response.read().decode('utf-8'))
                if 'name' in res_data:
                    self.session_name = res_data['name']
                    print(f"Successfully created Spanner TrueTime session: {self.session_name}", flush=True)
                else:
                    print(f"Session name not found in response: {res_data}", file=sys.stderr)
                    sys.exit(1)
        except Exception as e:
            print(f"Failed to initialize Spanner TrueTime session: {e}", file=sys.stderr)
            sys.exit(1)

    def next_id_string(self):
        if not self.session_name:
            print("Spanner TrueTime session not initialized.", file=sys.stderr)
            return ""

        # Use a lock to ensure only one thread uses the session at a time
        # The emulator might not support concurrent transactions on the same session well
        with self.lock:
            try:
                # 1. Begin Transaction
                begin_url = f"http://spanner:9020/v1/{self.session_name}:beginTransaction"
                begin_data = json.dumps({"options": {"readWrite": {}}}).encode('utf-8')
                begin_req = urllib.request.Request(begin_url, data=begin_data, headers={'Content-Type': 'application/json'}, method='POST')
                
                with urllib.request.urlopen(begin_req) as response:
                    begin_res = json.loads(response.read().decode('utf-8'))
                    transaction_id = begin_res.get('id')
                    
                if not transaction_id:
                    print("Failed to begin transaction", file=sys.stderr)
                    return ""

                # 2. Execute SQL
                url = f"http://spanner:9020/v1/{self.session_name}:executeSql"
                data = json.dumps({
                    "sql": "INSERT INTO events (description) VALUES ('event') THEN RETURN id",
                    "transaction": {"id": transaction_id}
                }).encode('utf-8')
                req = urllib.request.Request(url, data=data, headers={'Content-Type': 'application/json'}, method='POST')
                
                generated_id = ""
                with urllib.request.urlopen(req) as response:
                    res_data = json.loads(response.read().decode('utf-8'))
                    if 'rows' in res_data and len(res_data['rows']) > 0 and len(res_data['rows'][0]) > 0:
                        generated_id = str(res_data['rows'][0][0])
                    else:
                        print(f"Failed to extract ID from response: {res_data}", file=sys.stderr)

                # 3. Commit Transaction
                commit_url = f"http://spanner:9020/v1/{self.session_name}:commit"
                commit_data = json.dumps({
                    "transactionId": transaction_id,
                    "mutations": []
                }).encode('utf-8')
                commit_req = urllib.request.Request(commit_url, data=commit_data, headers={'Content-Type': 'application/json'}, method='POST')
                urllib.request.urlopen(commit_req)

                return generated_id
            except urllib.error.HTTPError as e:
                error_body = e.read().decode('utf-8')
                print(f"Error executing SQL: {e}. Body: {error_body}", file=sys.stderr)
                return ""
            except Exception as e:
                print(f"Error executing SQL: {e}", file=sys.stderr)
                return ""

File: src/node/generator/lib/spannertruetime/spanner_truetime.js
const http = require('http');

class SpannerTrueTimeGenerator {
    constructor() {
        this.sessionName = null;
    }

    async init() {
        for (let i = 0; i < 30; i++) {
            try {
                await this.makeRequest('http://spanner:9020/', 'GET');
                break;
            } catch (err) {
                console.log("Waiting for Spanner TrueTime emulator...");
                await new Promise(resolve => setTimeout(resolve, 2000));
            }
        }

        try {
            const dbName = "projects/test-project/instances/test-instance/databases/test-db";
            const response = await this.makeRequest(`http://spanner:9020/v1/${dbName}/sessions`, 'POST', {});
            const data = JSON.parse(response);
            if (data.name) {
                this.sessionName = data.name;
                console.log(`Successfully created Spanner TrueTime session: ${this.sessionName}`);
            } else {
                console.error("Session name not found in response:", data);
                process.exit(1);
            }
        } catch (err) {
            console.error("Failed to initialize Spanner TrueTime session:", err);
            process.exit(1);
        }
    }

    makeRequest(url, method, body = null) {
        return new Promise((resolve, reject) => {
            const { hostname, port, pathname } = new URL(url);
            const options = {
                hostname,
                port: port || 80,
                path: pathname,
                method,
                headers: {
                    'Content-Type': 'application/json'
                }
            };

            const req = http.request(options, (res) => {
                let data = '';
                res.on('data', (chunk) => {
                    data += chunk;
                });
                res.on('end', () => {
                    if (res.statusCode >= 200 && res.statusCode < 300) {
                        resolve(data);
                    } else if (res.statusCode === 404 && method === 'GET') {
                        resolve(data); // Emulator returns 404 for root GET but it means it's up
                    } else {
                        reject(new Error(`Request failed with status code ${res.statusCode}: ${data}`));
                    }
                });
            });

            req.on('error', (err) => {
                reject(err);
            });

            if (body) {
                req.write(JSON.stringify(body));
            }
            req.end();
        });
    }

    async nextIdString() {
        if (!this.sessionName) {
            console.error("Spanner TrueTime session not initialized.");
            return "";
        }

        try {
            // 1. Begin Transaction
            const beginTxUrl = `http://spanner:9020/v1/${this.sessionName}:beginTransaction`;
            const beginTxBody = {
                options: {
                    readWrite: {}
                }
            };
            const beginTxResponse = await this.makeRequest(beginTxUrl, 'POST', beginTxBody);
            const txData = JSON.parse(beginTxResponse);
            const transactionId = txData.id;

            if (!transactionId) {
                console.error("Failed to begin transaction:", txData);
                return "";
            }

            // 2. Execute SQL
            const url = `http://spanner:9020/v1/${this.sessionName}:executeSql`;
            const body = { 
                sql: "INSERT INTO events (description) VALUES ('event') THEN RETURN id",
                transaction: {
                    id: transactionId
                }
            };
            const response = await this.makeRequest(url, 'POST', body);
            const data = JSON.parse(response);

            let generatedId = "";
            if (data.rows && data.rows.length > 0 && data.rows[0].length > 0) {
                generatedId = data.rows[0][0];
            } else {
                console.error("Failed to extract ID from response:", data);
            }

            // 3. Commit Transaction
            const commitUrl = `http://spanner:9020/v1/${this.sessionName}:commit`;
            const commitBody = {
                transactionId: transactionId,
                mutations: []
            };
            await this.makeRequest(commitUrl, 'POST', commitBody);

            return generatedId;
        } catch (err) {
            console.error("Error executing SQL:", err);
            return "";
        }
    }
}

module.exports = SpannerTrueTimeGenerator;
